charms/istio-pilot/src/resources_handler.py

```diff
@@ -0,0 +1,152 @@
+            labels={"app.juju.is/created-by": f"{self.app_name}"}.update(labels),
+            namespace=namespace,
+        ):
+            self.delete_object(
+                obj,
+                namespace=namespace,
+                ignore_not_found=ignore_not_found,
+                ignore_unauthorized=ignore_unauthorized,
+            )
+
+    def apply_manifest(self, manifest, namespace=None):
+        for obj in codecs.load_all_yaml(manifest):
+            self.lightkube_client.apply(obj, namespace=namespace)
+
+    def delete_manifest(
+        self, manifest, namespace=None, ignore_not_found=False, ignore_unauthorized=False
+    ):
+        for obj in codecs.load_all_yaml(manifest):
+            self.delete_object(
+                obj,
+                namespace=namespace,
+                ignore_not_found=ignore_not_found,
+                ignore_unauthorized=ignore_unauthorized,
+            )
+
+    def create_ns_resource(self, filename: str, context: dict = None) -> GenericNamespacedResource:
+        """Returns a class representing a namespaced K8s resource.
+
+        Args:
+            - filename: name of the manifest file
+        """
+
+        # TODO: this is a generic context that is used for rendering
+        # the manifest files and extract their metadata. We should
+        # improve how we do this.
+        context = {
+            'namespace': 'namespace',
+            'app_name': 'name',
+            'name': 'generic_resource',
+            'request_headers': 'request_headers',
+            'response_headers': 'response_headers',
+            'port': 'port',
+            'service': 'service',
+        }
+
+        t = self.env.get_template(filename)
+        manifest = yaml.safe_load(t.render(context))
+        ns_resource = create_namespaced_resource(
+            group=manifest["apiVersion"].split("/")[0],
+            version=manifest["apiVersion"].split("/")[1],
+            kind=manifest["kind"],
+            plural=f"{manifest['kind']}s".lower(),
+            verbs=None,
+        )
+
+        return ns_resource
+
+    def reconcile_desired_resources(
+        self, resource: Any, desired_resources: str, namespace: str = None
+    ) -> None:
+        """Reconciles the desired list of resources of any kind.
+        Args:
+            resource: resource kind (e.g. Service, Pod)
+            desired_resource: all desired resources in manifest form as str
+            namespace (optional): namespace of the object
+        """
+
+        existing_resources = self.lightkube_client.list(
+            resource,
+            labels={
+                "app.juju.is/created-by": f"{self.app_name}",
+                f"app.{self.app_name}.io/is-workload-entity": "true",
+            },
+            namespace=namespace,
+        )
+        if desired_resources:
+            desired_resources = codecs.load_all_yaml(desired_resources)
+            diff_obj = set(existing_resources) - set(desired_resources)
```

Comment:
These objects are not hashable.  Might need to do something more nuanced to get a diff?  I'm seeing these in the debug-log:
```
Traceback (most recent call last):
  File "./src/charm.py", line 266, in <module>
    main(Operator)
  File "/var/lib/juju/agents/unit-istio-pilot-0/charm/venv/ops/main.py", line 424, in main
    framework.reemit()
  File "/var/lib/juju/agents/unit-istio-pilot-0/charm/venv/ops/framework.py", line 746, in reemit
    self._reemit()
  File "/var/lib/juju/agents/unit-istio-pilot-0/charm/venv/ops/framework.py", line 783, in _reemit
    custom_handler(event)
  File "./src/charm.py", line 177, in handle_ingress
    self._resource_handler.reconcile_desired_resources(
  File "/var/lib/juju/agents/unit-istio-pilot-0/charm/src/resources_handler.py", line 141, in reconcile_desired_resources
    diff_obj = set(existing_resources) - set(desired_resources)
TypeError: unhashable type: 'VirtualService'
```

Reply:
I think this is resolved with your latest changes.